ui, server: add remaining stats to Non-Table data

celiala opened the pull request:
Previously, the "Non-Table Cluster Data" table on the databases page showed
stats for just Time Series, but there are other types of non-table data that
weren't being shown.

This change add stats for the remaining non-table data to the "Non-Table
Cluster Data" table.

Things to verify for correctness / Outstanding Issues:

 1. There are four empty extra ranges for table descriptors that aren't actually tables, and I'm not sure how to reference via a span (`MetaRangesID, SystemRangesID, TimeseriesRangesID`, and `LivenessRangesID`). So the logic simply adds 4 ranges to `InternalUseStats.RangeCount`.

2. Even with the logic above to account for those 4 ranges, it looks like we're still missing 1 range. I've written up a test in hopes that someone might be able to help find this: https://github.com/celiala/cockroach/blob/display-non-table-non-ts-ranges/pkg/server/admin_test.go#L393

Notes:

 1. The user doesn't care what the other data types are, but simply that the
    stats are surfaced, so stats for all remaining non-table data types are
    stored into a single "Internal Use" bucket.

 2. It's also worth noting here that the sizes listed on the Databases page
    are RocksDB’s approximate sizes, so this number could be higher than
    expected when the data isn’t perfectly organized. To verify this, we
    checked ApproximateDiskBytes on startup:

```
--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -1309,6 +1309,12 @@ func (s *Server) Start(ctx context.Context) error {
        }
        s.stopper.AddCloser(&s.engines)

+       size, err := s.engines[0].ApproximateDiskBytes(roachpb.KeyMin, roachpb.KeyMax)
+       if err != nil {
+               return err
+       }
+       log.Warningf(ctx, "debug: engine claims there's %s in it", size)
+
        // Write listener info files early in the startup sequence. `listenerInfo` has a comment.
```

...against the size of the data dir:

```
./cockroach debug compact ./cockroach-data
du -sch ./cockroach-data
```

Fixes #24767

![image](https://user-images.githubusercontent.com/3051672/47459074-3432f600-d7a9-11e8-9f5f-284ce67b0999.png)

Release note (admin ui change): "Non-Table Cluster Data" on Databases page
now includes all non-table data types. Previously, this section only
showed Time Series data.

vilterp commented:
Looks good overall! Will follow up with a couple nits.

A couple notes on commit formatting:
- PR titles should include all top-level packages being modified (i.e. directories under `pkg/`). Went ahead and changed it here.
- Commits should do the same. I usually split server and UI work into separate commits (so one `ui:` and one `server:`), but I don't think it's a big deal to do so here.
- The last line of the commit message should be a release note, like this:
    ```
    ui: foo bar

    bla bla bla

    Release note(admin ui change): added foo bar
    ```
    There's supposed to be a git hook that makes `git commit` prompt you for this; not quite sure how to install it.

Also you have a couple of Go lint errors; I can show you how to view them in TeamCity. To see them locally run `make lintshort` (for Go errors) and `make ui-lint` (for UI errors).

vilterp commented on pkg/server/admin.go:
```diff
@@ -573,19 +573,53 @@ func (s *adminServer) TableStats(
+		// 		This won't work for [/Min, /Meta2), since this keyspace comes before Meta2 [and
+		// 		therefore won't have metadata RangeDescriptors in Meta2].
+		// 		We should probably figure out to get those stats (which may require something
+		// 		special).
```
I bet someone from Core can give us a trick to get stats for these ranges…

benesch replied:
I think it should just work? By my read the scan in `tableStatsForSpan(` doesn't just scan meta2; it should scan meta1, too, if you start your scan at KeyMin. Did you try this and run into trouble?

celiala replied:
Thanks for the hints!

I didn't get `LocalMax` to work as a start key, but was able to get span data for `[keys.Meta2Prefix, keys.System)` -- could you confirm that this is where I want to start scanning, as well as my understanding outlined below? Thanks!

Scans starting at `keys.KeyMin` and `keys.LocalMax` were both throwing error `pkg.storage.verifyKeys(L225): "start key in [/Min,/System) must be greater than LocalMax"`.

This happened because `keys.RangeMetaKey()` maps `keys.KeyMin`, `keys.LocalMax`, and `keys.Meta1Prefix` to the `/Min` range metadata key. _[But from reading the above, my understanding is that I don't want to scan any of this, due to issues with range-local addressing]._

The first `roachpb.Key` that maps beyond `/Min` range metadata key is `keys.Meta2Prefix`, which maps to `/Meta1` range metadata key. _[My understanding is that this is where I want to start scanning, since this is out of the range-local space]._

tschottdorf replied:
Hmm, `keys.LocalMax` *should* work. Maybe there's a buglet in the code you're calling? See here for an example that does include `LocalMax`:

https://github.com/cockroachdb/cockroach/blob/2b5799db2ddaac8211cd9b4128f951c8a4d71444/pkg/server/status_test.go#L500

This range-local stuff is extremely confusing and poorly designed/documented. If you have really understood it then you deserve a medal, because every time I run into it I have to spend a fair amount of time paging things back in.

The message you posted

> pkg.storage.verifyKeys(L225): "start key in [/Min,/System) must be greater than LocalMax".

indicates that you're scanning `[/Min,/System)`. I would check why that is. You should scan `[LocalMax, something)` as your "leftmost" request.

vilterp commented on pkg/server/admin.go:
```diff
@@ -573,19 +573,53 @@ func (s *adminServer) TableStats(
+				response.InternalUseStats.Stats.Add(
+					nonTableStats.Stats)
+				response.InternalUseStats.MissingNodes = append(
+					response.InternalUseStats.MissingNodes, nonTableStats.MissingNodes...)
```
Optional refactor: move this into a new `Add` method on `TableStatsResponse`:

```go
func (tsr *TableStatsResponse) Add(other *TableStatsResponse) {
```

Since this is a method on a struct generated from a proto, it should be defined in a file alongside the proto, with the same name. In this case, a new file `pkg/server/serverpb/admin.go`. The `Add` method on `MVCCStats` is a good example to follow: https://github.com/cockroachdb/cockroach/blob/d5bbbea83fb104eaa1051104b4ee941ccdf3a36f/pkg/storage/engine/enginepb/mvcc.go#L82

celiala replied:
ah, definitely -- done :)

vilterp commented on pkg/server/admin.go:
```diff
@@ -573,19 +573,53 @@ func (s *adminServer) TableStats(
+	response := serverpb.NonTableStatsResponse{}
+
+	for _, span := range nonTableSpans {
+		nonTableStats, err := s.tableStatsForSpan(ctx, span)
```
A bit odd that we're using a method called `tableStats` to get non-table stats. You could rename it to `spanStats` or something.

celiala replied:
did a half-solution: can't rename the response object but at least renamed the function name to avoid confusion for the reader (put refactor in separate commit).

couchand reviewed (approved):
:tada: congrats on your first PR @celia!  Modulo the lint issues @vilterp mentioned, it looks ready to go.  I did leave two code organization suggestions, let me know if you'd like to talk about either of them.

I would echo the points Pete made and observe that there's lots of useful content on the wiki, for instance https://github.com/cockroachdb/cockroach/wiki/Git-Commit-Messages.  The git hooks are in the repo in a folder called githooks.

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-:-LPbrn1K-ZqU6EphcpaD:b-2ndw4x)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/admin.go, line 607 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LPbqqk66vv-a1RaCZH6:-LPbqqk66vv-a1RaCZH7:b-48d524) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6d67c3ad36dbb369878ed9d2a20961a690c7ad/pkg/server/admin.go#L607)):*
> ```Go
> 			return nil, err
> 		}
> 		if span.Key.Equal(keys.TimeseriesPrefix) {
> ```

This seems a bit awkward to put the timeseries span in the list at all, only to have to special case it in the loop.
___
*[pkg/server/admin.go, line 613 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LPbqhZhDRULIyNjCZNh:-LPbqhZi9_28JFjymQ0B:b-j352jh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6d67c3ad36dbb369878ed9d2a20961a690c7ad/pkg/server/admin.go#L613)):*
> ```Go
> 				response.InternalUseStats = nonTableStats
> 			} else {
> 				response.InternalUseStats.RangeCount           += nonTableStats.RangeCount
> ```

I'd recommend pulling out the arithmetic so that it can be unit tested. 


<!-- Sent from Reviewable.io -->

vilterp commented on pkg/server/admin_test.go:
```diff
@@ -346,6 +346,46 @@ func TestAdminAPIDatabaseSQLInjection(t *testing.T) {
+	defer testCluster.Stopper().Stop(context.Background())
+	s := testCluster.Server(0)
+
+	// skipping TableStatsResponse.Stats comparison, since it includes data which aren't consistent (time, bytes)
```
Nit: phrase as a sentence (this is a convention we have)
```suggestion
	// Skip TableStatsResponse.Stats comparison, since it includes data which aren't consistent (time, bytes).
```

celiala replied:
oops, I didn't read convention before pushing PR. Read and fixed :)

couchand commented on pkg/server/serverpb/admin.proto:
```diff
@@ -227,6 +227,8 @@ message NonTableStatsRequest {
 message NonTableStatsResponse {
   // Information on time series ranges,
   TableStatsResponse time_series_stats = 1;
+  // Information for remaining non-table [non-time-series] ranges
```
or maybe "non-table, non-time-series"

couchand replied:
or maybe, since it's further clarifying remaining, "remaining (non-table, non-time-series) ranges"?

...maybe spending too much time thinking about the punctuation of a comment...

celiala committed:
server: rename tableStatsForSpan to statsForSpan

Release note: None

celiala committed:
server: add special case logic to scan range1

Release note: None

celiala committed:
ui, server: add remaining stats to Non-Table data

Previously, the "Non-Table Cluster Data" table on the databases page showed
stats for just Time Series, but there are other types of non-table data that
weren't being shown.

This change add stats for the remaining non-table data to the "Non-Table
Cluster Data" table.

Notes:

 1. The user doesn't care what the other data types are, but simply that the
    stats are surfaced, so stats for all remaining non-table data types are
    stored into a single "Internal Use" bucket.

 2. There are four empty extra ranges for table descriptors that aren't
    actually tables, and can't be referenced via a span (`MetaRangesID`,
    `SystemRangesID`, `TimeseriesRangesID`, and `LivenessRangesID`). For this
    case, logic simply adds 4 ranges to `InternalUseStats.RangeCount`.

 3. Even with the logic above to account the four ranges mentioned above, it
    looks like we're still missing 1 range (see
    `TestRangeCount_MissingOneRange`). This will be addressed separately as
    followup work, and has been tracked in #32099.

 4. It's also worth noting here that the sizes listed on the Databases page
    are RocksDB’s approximate sizes, so this number could be higher than
    expected when the data isn’t perfectly organized. To verify this, we
    checked ApproximateDiskBytes on startup:

```
--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -1309,6 +1309,12 @@ func (s *Server) Start(ctx context.Context) error {
        }
        s.stopper.AddCloser(&s.engines)

+       size, err := s.engines[0].ApproximateDiskBytes(roachpb.KeyMin, roachpb.KeyMax)
+       if err != nil {
+               return err
+       }
+       log.Warningf(ctx, "debug: engine claims there's %s in it", size)
+
        // Write listener info files early in the startup sequence. `listenerInfo` has a comment.
```

    ...against the size of the data dir:

```
./cockroach debug compact ./cockroach-data
du -sch ./cockroach-data
```

Fixes #24767

Release note (admin ui change): "Non-Table Cluster Data" on Databases page
now includes all non-table data types. Previously, this section only
showed Time Series data.

celiala commented on pkg/server/admin_test.go:
```diff
@@ -346,6 +349,111 @@ func TestAdminAPIDatabaseSQLInjection(t *testing.T) {
+// with no user data, all the ranges on the Databases page consist of:
+// 1) the total ranges listed for the system database
+// 2) the total ranges listed for the Non-Table data
+func TestRangeCount_MissingOneRange(t *testing.T) {
```
The ranges on the Databases page should now equal the total number of ranges on the Overview page. However, I'm still missing one.. ?

![image](https://user-images.githubusercontent.com/3051672/47819256-4cb68980-dd30-11e8-877a-14b6a51dc3c1.png)


![image](https://user-images.githubusercontent.com/3051672/47819223-3e686d80-dd30-11e8-9c3e-6a6d6428871d.png)

celiala commented:
PTAL, thanks!

- Non-Table now scans starting from LocalMax (thanks, Tobias!)
- Remaining PR suggestions have now been integrated
- There is still 1 range missing, though? (see my comment on PR)

couchand reviewed (changes requested):
Great!  Looks like there's two little lint failures in TeamCity on TODO style, and I noticed one copy-pasta comment.

As far as the last missing range, my strategy would be to enumerate them all through the range debug page (`/reports/range/:id`) and see what each of them says.  But looking into that can probably be addressed as followup work rather than blocking this PR; I've gone ahead and made #32099 to be sure we keep track of it.

> Reviewed 6 of 7 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-:-LQEjgY7-94Uf4-YfT3o:bmknotj)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/admin.go, line 69 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LQEkjzBAm1KcuteojBi:-LQEkjzCAMNL0zrFzAeF:be0j1ky) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7cb06344b50ba2251f7ed7a1b7267f7327d68c4/pkg/server/admin.go#L69)):*
> ```Go
> 	// e.g. descriptors 17, 18, 19, and 22 which correspond to MetaRangesID,
> 	// SystemRangesID, TimeseriesRangesID, and LivenessRangesID in pkg/keys.
> 	nonTableDescriptorRangeCount = 4
> ```

If we can't find that other range should we just mysteriously increment this?  half-joking, but only half.
___
*[pkg/server/admin_test.go, line 454 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LQElPSv-4CJzrX9Lz-0:-LQElPSv-4CJzrX9Lz-1:bi06jyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7cb06344b50ba2251f7ed7a1b7267f7327d68c4/pkg/server/admin_test.go#L454)):*
> ```Go
> 	assert.Equal(t,
> 		totalRangeCount,
> 		nonTableRangeCount+systemTableRangeCount+expectedMissingRangeCount)
> ```

I really like this test.  It's hard to imagine something more directly validating that we've addressed the original bug report.  (of course, the missing range is unnerving, but we'll figure that out I'm sure)
___
*[pkg/server/serverpb/admin.go, line 32 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LQEmN_2BUucR64x_RM0:-LQEmN_2BUucR64x_RM1:b-n8psux) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7cb06344b50ba2251f7ed7a1b7267f7327d68c4/pkg/server/serverpb/admin.go#L32)):*
> ```Go
> 	// and dropping subsequent ones seems reasonable to do, and is what is done
> 	// here.
> 	missingNodeIds := make(map[string]bool)
> ```

micro-optimization: you can use an empty struct as the map value type here to save one bit per node!
___
*[pkg/server/serverpb/admin_test.go, line 23 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LQEn6Pf6ytTHv-prjXc:-LQEn6PgEosBfoxChysT:byqsja4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7cb06344b50ba2251f7ed7a1b7267f7327d68c4/pkg/server/serverpb/admin_test.go#L23)):*
> ```Go
> )
> 
> // TestAdminDebugMetrics verifies that cmdline and memstats variables are
> ```

This comment seems to need a refresh...


<!-- Sent from Reviewable.io -->

celiala commented:
bors r+

celiala reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31830)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/admin.go, line 586 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LPbjtyQbj1Yf84MVlP9-r1-586:-LQFSzuX7r01Ns_P9N83:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6d67c3ad36dbb369878ed9d2a20961a690c7ad/pkg/server/admin.go#L586)):*
<details><summary><i>Previously, celiala wrote…</i></summary><blockquote>

Tobias and I paired on this and got this to work! (Tobias' fix is incorporated into this PR).
</blockquote></details>

Done.
___
*[pkg/server/admin.go, line 603 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LPbjtyQbj1Yf84MVlP9-r1-603:-LQFU0x8B_hRFKQ4q8Wl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6d67c3ad36dbb369878ed9d2a20961a690c7ad/pkg/server/admin.go#L603)):*
<details><summary><i>Previously, celiala wrote…</i></summary><blockquote>

did a half-solution: can't rename the response object but at least renamed the function name to avoid confusion for the reader (put refactor in separate commit).
</blockquote></details>

Done.
___
*[pkg/server/admin.go, line 620 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LPbjtyQbj1Yf84MVlP9-r1-620:-LQFU7UVB6Qx4_NoJQC6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6d67c3ad36dbb369878ed9d2a20961a690c7ad/pkg/server/admin.go#L620)):*
<details><summary><i>Previously, celiala wrote…</i></summary><blockquote>

ah, definitely -- done :)
</blockquote></details>

Done.
___
*[pkg/server/admin.go, line 69 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LQEkjzBAm1KcuteojBi:-LQFTGGs22Sazy9dB0se:bvz2sal) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7cb06344b50ba2251f7ed7a1b7267f7327d68c4/pkg/server/admin.go#L69)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

If we can't find that other range should we just mysteriously increment this?  half-joking, but only half.
</blockquote></details>

This may what ends up happening, as a result of looking into #32099.
___
*[pkg/server/admin_test.go, line 382 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LPbjtyQbj1Yf84MVlPA-r1-382:-LQFUFB9CHl8jwwmpTYK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6d67c3ad36dbb369878ed9d2a20961a690c7ad/pkg/server/admin_test.go#L382)):*
<details><summary><i>Previously, celiala wrote…</i></summary><blockquote>

thanks for the pointer - done :)
</blockquote></details>

Done.
___
*[pkg/server/serverpb/admin_test.go, line 23 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-LQEn6Pf6ytTHv-prjXc:-LQFZtb5ACFhWCzY9Uak:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7cb06344b50ba2251f7ed7a1b7267f7327d68c4/pkg/server/serverpb/admin_test.go#L23)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

This comment seems to need a refresh...
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

celiala commented:
Resolved feedback. PTAL - thanks!

couchand reviewed (approved):
Good to go!

> Reviewed 7 of 7 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31830#-:-LQp7fXH2eZV3UG69XSZ:bf7qwsy)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

celiala commented:
bors r+
